Improved id resolving of entities

iltar opened the pull request:
Fixes #12.

The `EntityIdResolver` now uses the by doctrine defined field which uses `@Id`. It also supports relations which are a PK, it will recursively try to resolve it: 
 - Client 1..1 Authentication
 -  `authentication.client.id` will be called by using the property Accessor. when you pass the authentication entity

There's a limitation though, when you have a composite primary key, it cannot be resolved to a single scalar.

yannickl88 commented on test/Functional/Fixtures/Entity/Authentication.php:
```diff
@@ -0,0 +1,27 @@
+     */
+    private $client;
+
+    public function __construct(Client $client)
```
missing doc

WouterJ commented on src/DoctrineBridge/RecursiveIdFinder.php:
```diff
@@ -0,0 +1,58 @@
+            throw new CompositePrimaryKeyException(sprintf('Composite Primary Keys cannot be resolved to a single scalar value on Entity: %s', get_class($entity)));
+        }
+
+        $pkValue = $this->propertyAccessor->getValue($entity, $identifier[0]);
```
IIRC, Doctrine uses reflection, it may be possible that the property accessor cannot access the ID property. I'm not sure if you want to change it though

iltar replied:
If the property is private with no getter, it does not want to be exposed so I believe that using reflection would violate this abstraction

xabbuh replied:
But you may want to handle that case with a dedicated exception then to ease debugging.

WouterJ commented on src/DoctrineBridge/Router/EntityIdResolver.php:
```diff
@@ -4,7 +4,7 @@
 use Doctrine\Common\Persistence\ManagerRegistry;
 use Doctrine\ORM\Mapping\MappingException;
-use Iltar\HttpBundle\Exception\UncallableMethodException;
+use Iltar\HttpBundle\DoctrineBridge\IdFromEntityInterface;
```
I would expect this interface to not live in DoctrineBridge, as the goal of the interface is to allow other implementations (which don't depend on Doctrine).

iltar replied:
Indeed, I will move this. This can be re-used easily by Eloquent, propel or any custom ORM which uses an id. 

Now that I think of it, I could move the whole IdResolver out of the bridge and let the ORM specific implementation handle the id resolving. I should probably add an `isManaged($entity)` to the `IdFromEntityInterface`. In this case I could also call it `ManagedModelInterface` or something, any suggestions on that?

WouterJ commented on src/Resources/config/entity_id_resolver.xml:
```diff
@@ -9,7 +9,12 @@
+            <argument type="service" id="iltar_http.recursive_id_finder" />
             <tag name="router.parameter_resolver" priority="100" />
         </service>
+        <service id="iltar_http.recursive_id_finder" class="Iltar\HttpBundle\DoctrineBridge\RecursiveIdFinder">
```
can be private

WouterJ commented on test/autoload.php:
```diff
@@ -0,0 +1,13 @@
+<?php
```
you won't need this file if you require `symfony/phpunit-bridge`.

iltar replied:
Ah right, completely forgot about that

xabbuh commented on src/DoctrineBridge/Router/EntityIdResolver.php:
```diff
@@ -25,11 +25,18 @@
 
     /**
-     * @param ManagerRegistry $manager
+     * @var IdFromEntityInterface
```
Actually, this docblock is not needed at all as IDEs can infer the type from the constructor.

iltar replied:
For now I will leave them in case I need to add scalar values. Once I can use proper typehints, I will omit docs were possible. I'm thinking of making 2.0 only available for php 7.0.

xabbuh commented on src/DoctrineBridge/Router/EntityIdResolver.php:
```diff
@@ -25,11 +25,18 @@
+    /**
+     * @param ManagerRegistry       $manager
+     * @param IdFromEntityInterface $idFromEntity
+     */
```
same here (type information can be deduced from the type hints)

iltar commented:
@WouterJ besides of all the docs/deprecation triggers to be added/config and tests, do you think this is in a more extendable setup?
 - Implement the `ModelDescriptorInterface` when resolving is needed for like an entity/model or what so ever
 - All logic from the previous entity id resolver is now in the `EntityIdDescriptor`
 - I will add (proper) config to configure the model descriptors in the future, not in this PR due to the size

Any idea why the phpunit-bridge is not doing what it should on travis? Locally it works fine.

WouterJ commented:
> Any idea why the phpunit-bridge is not doing what it should on travis? Locally it works fine. was a cache locally, broken too - missing the actual dependency.

Looks like it does what it should do?

iltar commented:
@WouterJ Yeah, now it does. I forgot to require it, thought I already had it. So when I ran it locally, the cache made it succeed for some reason. 3 `--amend` later and it worked.

iltar commented:
Okay so I've added tests and modified the readme some.

In a future patch I will deprecate the config `entity_id_resolver` and instead make a dynamic config that allows you to register `ModelDescriptorInterface` implementations.

@WouterJ was this more what you had in mind regarding the re-usability?

WouterJ commented on .travis.yml:
```diff
@@ -18,7 +21,7 @@ matrix:
       env: COMPOSER_FLAGS="--prefer-lowest"
     - php: 7.0
-      env: COMPOSER_FLAGS="" SYMFONY_VERSION="3.0.*"
+      env: COMPOSER_FLAGS="" SYMFONY_VERSION="3.1.*"
```
`COMPOSER_FLAGS=""` can be removed and a job for sf 3.0 should be added.

iltar replied:
How should I define this?

WouterJ commented on src/DoctrineBridge/Router/EntityIdResolver.php:
```diff
@@ -2,73 +2,17 @@
-use Iltar\HttpBundle\Router\ParameterResolverInterface;
+use Iltar\HttpBundle\Router\Resolver\IdentifyingValueResolver;
+
+@trigger_error(sprintf('%s is deprecated as of 1.1 and will be removed in 2.0. Use the %s instead.', EntityIdResolver::class, IdentifyingValueResolver::class), E_USER_DEPRECATED);
```
should be put in the constructor

WouterJ commented on src/DoctrineBridge/Router/EntityIdResolver.php:
```diff
@@ -2,73 +2,17 @@
- * A resolver using the ManagerRegistry to call getId if possible.
+ * Uses a model descriptor to get the identifying value of a model.
  *
+ * @deprecated as of 1.1, please use ... instead
```
replace ... with IdentifyingValueResolver

iltar commented:
@WouterJ something like this?
